### PR TITLE
[REEF-1405] Cache the data in IMRU Context layer

### PR DESCRIPTION
**Pull request by jwang98052**
* Download data file in DataLoadingContext when context starts
* Add test cases to inject FilePartitionDataSet to IMRUDriver.
* Verify data file has been downloaded in mapper function in test

JIRA: [REEF-1405](https://issues.apache.org/jira/browse/REEF-1405)
This closes

### CODE REVIEW

**Comment by tcNickolas** [comment]
Looking

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+                        .NewInjector(partitionDescriptor.GetPartitionConfiguration())
+                        .GetInstance<IInputPartition<IEnumerable<Row>>>();
+
+                using (partition as IDisposable)
+                {
+                    IEnumerable<Row> e = partition.GetPartitionHandle();
+
+                    foreach (var row in e)
+                    {
+                        Logger.Log(Level.Info, string.Format(CultureInfo.CurrentCulture, "Data read {0}: ", row.GetValue()));
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Please use shorter `Logger.Log(Level.Info, "Data read {0}: ", row.GetValue());` instead

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                updateTaskMemory,
+                testFolder);
+            ValidateSuccessForLocalRuntime(numTasks, 0, 0, testFolder);
+            CleanUp(testFolder);
+        }
+
+        /// <summary>
+        /// This method overrides base class method to pass IEnumerable<Row> as TPartitionType
+        /// </summary>
+        /// <param name="runOnYarn"></param>
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Can we delete these empty tags in new methods if we have no plans to fill them? Same for `DriverEventHandlerConfigurations` method

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            string testFolder = DefaultRuntimeFolder + TestId;
+            TestBroadCastAndReduce(false,
+                numTasks,
+                chunkSize,
+                dims,
+                iterations,
+                mapperMemory,
+                updateTaskMemory,
+                testFolder);
+            ValidateSuccessForLocalRuntime(numTasks, 0, 0, testFolder);
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: For verification we rely on the fact that if files are not copied, the tasks will fail to be constructed, and we'll get non-zero number of failed tasks, right? Might be worth to leave a comment about this in code, since for IMRU tests we tend to to more validation than just `ValidateSuccessForLocalRuntime`, so having no extra validation is a bit confusing

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            private int _iterations;
+
+            [Inject]
+            private TestSenderMapFunction(
+                [Parameter(typeof(TaskConfigurationOptions.Identifier))] string taskId,
+                IInputPartition<T> partition)
+            {
+                var tempFile = TangFactory.GetTang().NewInjector().GetNamedInstance<TempFileFolder, string>();
+                var tmpFileFodler = Directory.GetCurrentDirectory() + tempFile;
+                Assert.True(Directory.Exists(tmpFileFodler));
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Should we also verify here that there is exactly one file in this folder, and it has a name which matches mask "8 random characters 0-9|a-f"? Theoretically if something goes wrong there can be no files or too many files in this folder

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        /// Create test data as bytes
+        /// </summary>
+        /// <param name="dataNumber"></param>
+        /// <returns></returns>
+        private static byte[] CreateTestData(int dataNumber)
+        {
+            var bytes = new byte[dataNumber];
+            for (int i = 0; i < dataNumber; i++)
+            {
+                bytes[i] = (byte)i;
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: A little readability improvement: if we write `(byte)(i + 'a')` here, test data generated will be readable characters. I think it's better to work with human-readable data if possible

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        }
+
+        /// <summary>
+        /// Build RowSerialzer configuration and serialize it as string so that to pass it to driver
+        /// </summary>
+        /// <returns></returns>
+        private static string GetRowSerializerConfigString()
+        {
+            var serializerConf = TangFactory.GetTang().NewConfigurationBuilder()
+                .BindImplementation<IFileDeSerializer<IEnumerable<Row>>, RowSerializer>(
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: `<IFileDeSerializer<IEnumerable<Row>>, RowSerializer>` can be omitted (we use `BindImplementation` without `<>` more frequently).

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+                    GenericType<IFileDeSerializer<IEnumerable<Row>>>.Class,
+                    GenericType<RowSerializer>.Class)
+                .Build();
+            return (new AvroConfigurationSerializer()).ToString(serializerConf);
+        }
+
+        /// <summary>
+        /// Test Row class
+        /// </summary>
+        public class Row
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Do we plan to share this class with other tests? If no, it can be made private

**Comment by tcNickolas** [comment]
I've made a pass

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 7 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Done

**Comment by jwang98052** [comment]
@tcNickolas I have addressed your review comments. Please have another look. 
@dkm2110 , could you please also review it? Thanks!

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 8]
```diff
@@ -74,20 +77,12 @@ public void TestWithFilePartitionDataSetOnLocalRuntime()
                 updateTaskMemory,
                 testFolder);
             ValidateSuccessForLocalRuntime(numTasks, 0, 0, testFolder);
-            CleanUp(testFolder);
+            ////CleanUp(testFolder);
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: For the final version of the test, you probably want to uncomment this line

**Comment by tcNickolas** [comment]
I left two minor comments, other than that, LGTM. Will wait for @dkm2110's review

**Comment by dkm2110** [comment]
@jwang98052  looked at PR. One comment: looking at FileBasedInputPartition, it should now create instance of T  and store it in a local variable when Cache is called. Shouldn't that change also be part of this PR or you want to do it as a separate one? Otherwise LGTM.

**Comment by jwang98052** [comment]
@dkm2110 do you mean this one? _data here is a local variable that holds T. 
private Optional<T> _data = Optional<T>.Empty();
 _data = Optional<T>.Of(_fileSerializer.Deserialize(_localFiles.Value));

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+                Assert.True(directory.Contains("-partition-"));
+
+                var files = Directory.EnumerateFiles(directory);
+                Assert.Equal(1, files.Count());
+                var file = files.FirstOrDefault();
+                var a = file.Split('\\');
+                var fileName = a[a.Length - 1];
+                Assert.Equal(8, fileName.Length);
+
+                var matchCounter = Regex.Matches("40af4c53", @"[a-zA-Z0-9]").Count;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: Yes. It was my test data :).

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithFilePartitionDataSetTest.cs]: sure.

**Comment by dkm2110** [comment]
@jwang98052 yes thats what I am talking about.

**Comment by jwang98052** [comment]
@tcNickolas I have addressed your comment. 
@dkm2110 the change is in the 3rd commit. Please review it. Thanks!

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs [hunk 10]
```diff
@@ -252,7 +252,7 @@ protected void PeriodicUploadLog(object source, ElapsedEventArgs e)
             }
         }
 
-        internal string[] ReadLogFile(string logFileName, string subfolder = "driver", string testFolder = DefaultRuntimeFolder, int retryCount = 60)
+        internal string[] ReadLogFile(string logFileName, string subfolder = "driver", string testFolder = DefaultRuntimeFolder, int retryCount = 100)
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: Why is this change needed?

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    public class IMRUBroadcastReduceWithLocalFile : IMRUBroadcastReduceWithFilePartitionDataSetTest
+    {
+        /// <summary>
+        /// This test tests DataLoadingContext with FilePartitionDataSet
+        /// The IPartitionedInputDataSet configured in this test is passed to IMRUDriver, and IInputPartition associated with it is injected in 
+        /// DataLoadingContext for the Evaluator. Cache method in IInputPartition is called when the context starts. 
+        /// The local file is used in the test, so CopyToLocal will be set to false. The validation will be inside test Mapper function.
+        /// </summary>
+        [Fact]
+        public void TestWithFilePartitionDataSetForLocalFile()
```
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs]: What is the main difference from IMRUBroadcastReduceWithFilePartitionDataSetTest ? Is it mainly that copy to local is set to false? I am wondering if you combine the two by passing extra boolean parameter in map function?

**Comment by dkm2110** [comment]
@jwang98052  left  comments.

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs (hunk 11 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs]: done

File: lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs (hunk 10 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs]: broadcast tests usually takes longer time for the log files to be completed. I will  change it by passing parameters.

**Comment by jwang98052** [comment]
@dkm2110 I have addressed your  comments.

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs (hunk 11 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs]: Should this now be a separate class in itself or this test be moved to previous class above?

**Comment by dkm2110** [comment]
@jwang98052  left one minor comment. Otherwise LGTM.
I do not think I need to look at it again.

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs (hunk 11 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBroadcastReduceWithLocalFile.cs]: BuildPartitionedDatasetConfiguration() is an overriding method and the signature cannot be changed, that means I have no way to pass a bool to the method in the same test. Therefore, using a subclass would allow to override this method to provide different named parameter value.

**Comment by tcNickolas** [comment]
@jwang98052 Looks like all comments are addressed. I'll test and merge
